server/app/durablejobs/jobs/AddCategoryAndTranslationsJob.java

```diff
@@ -0,0 +1,111 @@
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import controllers.dev.seeding.CategoryTranslationFileParser;
+import durablejobs.DurableJob;
+import io.ebean.DB;
+import io.ebean.Database;
+import io.ebean.Transaction;
+import io.ebean.TxScope;
+import java.util.List;
+import models.CategoryModel;
+import models.PersistedDurableJobModel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import play.Environment;
+import repository.CategoryRepository;
+import services.LocalizedStrings;
+
+/* Iterates through all categories and ensures translations aren't missing. */
+public final class AddCategoryAndTranslationsJob extends DurableJob {
+  private static final Logger LOGGER = LoggerFactory.getLogger(AddCategoryAndTranslationsJob.class);
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+
+  private final CategoryRepository categoryRepository;
+  private final Environment environment;
+  private final PersistedDurableJobModel persistedDurableJobModel;
+  private final Database database;
+
+  public AddCategoryAndTranslationsJob(
+      CategoryRepository categoryRepository,
+      Environment environment,
+      PersistedDurableJobModel persistedDurableJobModel) {
+    this.categoryRepository = checkNotNull(categoryRepository);
+    this.environment = checkNotNull(environment);
+    this.persistedDurableJobModel = persistedDurableJobModel;
+    this.database = DB.getDefault();
+  }
+
+  @Override
+  public PersistedDurableJobModel getPersistedDurableJob() {
+    return persistedDurableJobModel;
+  }
+
+  @Override
+  public void run() {
+    LOGGER.info("Starting job to update categories with translations.");
+    int errorCount = 0;
+    try (Transaction jobTransaction = database.beginTransaction()) {
+      try {
+        // Parse messages files for most up to date translations
+        CategoryTranslationFileParser messagesFileParser =
+            new CategoryTranslationFileParser(environment);
+        List<CategoryModel> categoriesFromMessagesFile =
+            messagesFileParser.createCategoryModelList();
+
+        // Iterate through categories from the messages files and compare with categories in the DB
+        for (CategoryModel messageFileCategory : categoriesFromMessagesFile) {
+          // Get the database category or create one if there is none matching the category from the
+          // messages file
+          CategoryModel dbCategory =
+              categoryRepository.fetchOrSaveUniqueCategory(messageFileCategory);
+          LocalizedStrings dbTranslations = dbCategory.getLocalizedName();
+          LocalizedStrings fileTranslations = messageFileCategory.getLocalizedName();
+          // Update the database if the translations in the messages file differs
+          if (!dbTranslations.equals(fileTranslations)) {
```

Comment:
**question:** Since this is using the default `Object.equals`, won't `!dbTranslations.equals` always return true, even if the translations are the same, since `dbTranslations` and `fileTranslations` point to two different addresses in memory?

I've just seen that you did check that no updates are made when it isn't necessary in the `run_makesNoUpdatesIfCategoriesMatch` test.  Could you please help me understand how that's working with the default `equals`?

Reply:
Good question! Its kind of sneaky but since LocalizedStrings uses AutoValue, it uses their override of equals, which checks for the equality of the data, not the addresses in memory. Section 3 in [these docs](https://www.baeldung.com/introduction-to-autovalue) give an explanation of it, but let me know if you have questions.